server/src/server.rs

```diff
@@ -652,20 +663,34 @@ impl Server {
     }
 
     async fn check_resume_forwarding(&self) {
-        let mut server = self.state.write().await;
         let clock = DefaultClock::default();
-        let rs_fwd = server.resume_forwarding.clone();
+        let rs_fwd = {
+            let server = self.state.read().await;
+            server.resume_forwarding.clone()
```

Comment:
Here you may already take the relevant items from the collection. You wouldn't have to remove `elem`s in the loop below and the lock would become a `read`